### PR TITLE
V8

### PR DESCRIPTION
**Pull request by zrlk**

### CODE REVIEW

File: anodyne/tools/BUILD [hunk 1]
```diff
@@ -0,0 +1,22 @@
+load(":v8_heap_gen.bzl", "v8_heap_gen")
+
+cc_binary(
+    name = "v8_heap_gen",
+    srcs = ["v8_heap_gen.cc"],
+    deps = ["//third_party/v8"],
+)
+
+v8_heap_gen(
+    "base_heap",
```
Comment (creachadair) [inline anodyne/tools/BUILD]: Please use `name = value` format and omit the default values.

File: anodyne/tools/v8_heap_gen.bzl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Build a V8 heap snapshot from the JavaScript files in srcs, emitting
+# out_prefix.h and out_prefix.cc, which provide the function
+# `::InitV8FromBuiltinBuffers()` and the global
+# `v8::StartupData kIsolateInitBlob`.
+def v8_heap_gen(name, srcs, out_prefix):
```
Comment (creachadair) [inline anodyne/tools/v8_heap_gen.bzl]: `def v8_heap_gen(name, srcs=[], out_prefix=''):`

You can check the values and use `fail` to report an error inside the macro body, I think that's probably better than relying on positional arguments even though they do work.

File: third_party/v8/BUILD [hunk 3]
```diff
@@ -0,0 +1,33 @@
+licenses(["notice"])
+
+package(default_visibility = ["//visibility:public"])
```
Comment (creachadair) [inline third_party/v8/BUILD]: Why public?

Could we leave this to the default and only export the images that are needed?

File: anodyne/tools/v8_heap_gen.cc [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+// and contain the isolate heap resulting from image.js (respectively).
+
+namespace {
+void GetFileContent(const char* filename, std::string* out) {
+  ::FILE* f = ::fopen(filename, "rb");
+  if (!f) {
+    ::fprintf(stderr, "can't open %s\n", filename);
+    exit(1);
+  }
+  if (::fseek(f, 0, SEEK_END)) {
```
Comment (creachadair) [inline anodyne/tools/v8_heap_gen.cc]: Would `fstat(2)` work here?

File: anodyne/tools/v8_heap_gen.cc [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    exit(1);
+  }
+  out->resize(size);
+  if (::fread(const_cast<char*>(out->data()), size, 1, f) != 1) {
+    ::fprintf(stderr, "can't read %s\n", filename);
+    exit(1);
+  }
+  ::fclose(f);
+}
+void EmitData(FILE* f, const std::string& name, const char* data, size_t len) {
```
Comment (creachadair) [inline anodyne/tools/v8_heap_gen.cc]: Please leave blanks between the functions, and consider adding brief doc comments.

(I realize they're not exposed to callers, but it's still helpful for someone reading the binary)

File: third_party/v8/BUILD [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+exports_files([
+    "snapshot/snapshot_blob.bin",
+    "snapshot/natives_blob.bin",
+])
+
+cc_library(
+    name = "v8",
+    # Do not sort; srcs are link-order dependent.
```
Comment (creachadair) [inline third_party/v8/BUILD]: Does buildifier promise not to do that without an annotation?

File: README.md [hunk 7]
```diff
@@ -5,3 +5,13 @@ NOTE: This is not an official Google product.
 Anodyne is an open-source experimental project to implement a type analysis tool
 for commonly used manifestly-typed imperative programming languages.
 
+## Building
+
+Anodyne can be built with Bazel. There are some additional dependencies that
+are not managed by Bazel that you must provide. They may be hard- or soft-linked
+to these locations:
+
+  * v8 (17a6ec1b88f39beaa24a31c132ee27bc024c2d6b via [http://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/depot_tools_tutorial.html#_setting_up](depot_tools))
```
Comment (creachadair) [inline README.md]: I think these are reversed, viz., I believe you want 

```
  [`depot_tools`](http://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/depot_tools_tutorial.html#_setting_up)
```

Alternatively, GitHub flavored markdown supports forward refs:

```
  [`depot_tools`][dt]

[dt]: http://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/depot_tools_tutorial.html#_setting_up
```

File: anodyne/tools/BUILD (hunk 1 above)
Comment (zrlk) [reply anodyne/tools/BUILD]: Done

File: anodyne/tools/v8_heap_gen.bzl (hunk 2 above)
Comment (zrlk) [reply anodyne/tools/v8_heap_gen.bzl]: Done

File: third_party/v8/BUILD (hunk 3 above)
Comment (zrlk) [reply third_party/v8/BUILD]: We need the V8 API to actually use the images.

File: third_party/v8/BUILD (hunk 6 above)
Comment (zrlk) [reply third_party/v8/BUILD]: It does not.

File: third_party/v8/BUILD (hunk 3 above)
Comment (creachadair) [reply third_party/v8/BUILD]: Ah, I see. Can you please add a comment to that effect?

File: third_party/v8/BUILD (hunk 6 above)
Comment (creachadair) [reply third_party/v8/BUILD]: Oh, actually it looks like it recognized `# Do not sort` at the beginning of the comment, so I think we're OK here.

File: anodyne/tools/v8_heap_gen.cc (hunk 4 above)
Comment (zrlk) [reply anodyne/tools/v8_heap_gen.cc]: Done.

File: anodyne/tools/v8_heap_gen.cc (hunk 5 above)
Comment (zrlk) [reply anodyne/tools/v8_heap_gen.cc]: Done.

File: third_party/v8/BUILD (hunk 3 above)
Comment (zrlk) [reply third_party/v8/BUILD]: Done.

File: third_party/v8/BUILD (hunk 6 above)
Comment (zrlk) [reply third_party/v8/BUILD]: Yes, the "do not sort" comment is the annotation.

**Pull request merged**